### PR TITLE
#88 - User profile system

### LINKED ISSUE #88: User profile system
The user should be able to view and edit his profile.

- Upload and change avatar
- Change password
- Change username
- Change email?

### PR DESCRIPTION
**Pull request by Majkulnn**
user can change email, name, avatar and password

This should close #88

### CODE REVIEW

**Review by krzysztofrewak — APPROVED** [review]

File: resources/views/partials/navbar.blade.php [hunk 1]
```diff
@@ -58,14 +58,14 @@ class="bg-gray-800 flex text-sm rounded-full focus:outline-none focus:ring-2 foc
...
-                  src="https://images.unsplash.com/photo-1472099645785-5658abf4ff4e?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=facearea&facepad=2&w=256&h=256&q=80"
+                  src="{{Auth::user()->avatarPath}}"
                   alt="">
               </button>
             </div>
             <div x-show="userDropdownOpened" x-cloak x-transition
               class="origin-top-right absolute right-0 mt-2 w-48 rounded-md shadow-lg py-1 bg-white ring-1 ring-black ring-opacity-5 focus:outline-none z-10"
               id="user-menu" role="menu" aria-orientation="vertical" aria-labelledby="user-menu-button" tabindex="-1">
-              <a href="#" class="block px-4 py-2 text-sm text-gray-700" role="menuitem" tabindex="-1"
+              <a href="{{route('user.profile')}}" class="block px-4 py-2 text-sm text-gray-700" role="menuitem" tabindex="-1"
```
Comment (MichalMyskow) [inline resources/views/partials/navbar.blade.php:68]:
```suggestion
              <a href="{{ route('user.profile') }}" class="block px-4 py-2 text-sm text-gray-700" role="menuitem" tabindex="-1"
```
Same as above.

File: resources/views/partials/navbar.blade.php [hunk 2]
```diff
@@ -58,14 +58,14 @@ class="bg-gray-800 flex text-sm rounded-full focus:outline-none focus:ring-2 foc
                 id="user-menu-button" aria-expanded="false" aria-haspopup="true">
                 <span class="sr-only">Open user menu</span>
                 <img class="h-8 w-8 rounded-full"
-                  src="https://images.unsplash.com/photo-1472099645785-5658abf4ff4e?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=facearea&facepad=2&w=256&h=256&q=80"
+                  src="{{Auth::user()->avatarPath}}"
```
Comment (MichalMyskow) [inline resources/views/partials/navbar.blade.php:61]:
```suggestion
                  src="{{ Auth::user()->avatarPath }}"
```
Stay with convention

File: resources/views/user/profile/index.blade.php [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+                                   class="appearance-none block w-full px-3 py-2 border border-gray-300 rounded-md shadow-sm placeholder-gray-400 focus:outline-none focus:ring-indigo-500 focus:border-indigo-500 sm:text-sm" />
+                        </div>
+                    </div>
+
+                    <div>
+                        <label for="email" class="block text-sm font-medium text-gray-700">
+                            Email address
+                        </label>
+                        <div class="mt-1">
+                            <input id="email" name="email" readonly type="email" value="{{$user->email}}"
```
Comment (MichalMyskow) [inline resources/views/user/profile/index.blade.php:36]:
```suggestion
                            <input id="email" name="email" readonly type="email" value="{{ $user->email }}"
```

File: resources/views/user/profile/index.blade.php [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+                                   class="appearance-none block w-full px-3 py-2 border border-gray-300 rounded-md shadow-sm placeholder-gray-400 focus:outline-none focus:ring-indigo-500 focus:border-indigo-500 sm:text-sm" />
+                        </div>
+                    </div>
+                    <div>
+                        <a href="{{route("user.profile.edit")}}">
+                        <button type="submit"
+                                class="w-full flex justify-center py-2 px-4 border border-transparent rounded-md shadow-sm text-sm font-medium text-white bg-indigo-600 hover:bg-indigo-700 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-indigo-500">
+                            Edit profile
+                        </button></a>
+                        <a href="{{route("user.profile.password")}}">
```
Comment (MichalMyskow) [inline resources/views/user/profile/index.blade.php:46]:
```suggestion
                        <a href="{{ route("user.profile.edit") }}">
                        <button type="submit"
                                class="w-full flex justify-center py-2 px-4 border border-transparent rounded-md shadow-sm text-sm font-medium text-white bg-indigo-600 hover:bg-indigo-700 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-indigo-500">
                            Edit profile
                        </button></a>
                        <a href="{{ route("user.profile.password") }}">
```

File: src/Http/Controllers/AccountController.php [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+
+class AccountController extends Controller
+{
+    public function __construct(
+        protected  Guard $auth,
+    ) {}
+
+    public function index(): View
+    {
+        return view("user.profile.index")->with(["user" => $this->auth->user()]);
```
Comment (MichalMyskow) [inline src/Http/Controllers/AccountController.php:25]:
```suggestion
        return view("user.profile.index")
            ->with(["user" => $this->auth->user()]);
```

File: src/Http/Requests/Account/UpdateUserDataRequest.php [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+use Illuminate\Support\Facades\Auth;
+use Illuminate\Validation\Rule;
+
+class UpdateUserDataRequest extends FormRequest
+{
+    public function rules(): array
+    {
+        return [
+            "email" => ["required", "string", "email", "max:255",
+                Rule::unique("users")->where(fn ($query) => $query->where("email", "!=", Auth::user()->email)), ],
```
Comment (MichalMyskow) [inline src/Http/Requests/Account/UpdateUserDataRequest.php:18]: Please type function's returnable and parameters.

File: src/Models/User.php [hunk 7]
```diff
@@ -53,6 +55,10 @@ class User extends Model implements AuthenticatableContract, AuthorizableContrac
         "name",
         "email",
         "password",
+        "avatar_path",
+    ];
+    protected $attributes = [
+        "avatar_path" => Constants::SPEAKER_DEFAULT_AVATAR_PATH,
```
Comment (MichalMyskow) [inline src/Models/User.php:61]:
```suggestion
        "avatar_path" => Constants::USER_DEFAULT_AVATAR_PATH,
```
I think it should be User avatar

File: src/Services/UserRegisterService.php [hunk 8]
```diff
@@ -22,6 +23,7 @@ public function register(string $email, string $name, string $password): void
             "email" => $email,
             "name" => $name,
             "password" => $hashedPassword,
+            "avatar" => Constants::USER_DEFAULT_AVATAR_PATH,
```
Comment (MichalMyskow) [inline src/Services/UserRegisterService.php:26]:
```suggestion

```
You don't need that here. Default avatar is generated in the model.

**Review by MichalMyskow — CHANGES_REQUESTED** [review]

File: database/seeders/DummyDataSeeder.php [hunk 9]
```diff
@@ -22,6 +23,7 @@ public function run(): void
             "email" => "admin@example.com",
             "password" => Hash::make("password"),
             "email_verified_at" => Carbon::createFromDate(2022, 01, 01),
+            "avatar_path" => Constants::USER_DEFAULT_AVATAR_PATH,
```
Comment (MichalMyskow) [inline database/seeders/DummyDataSeeder.php:26]:
```suggestion
```
You don't need that here. Default avatar is generated in the model.

**Review by MichalMyskow — CHANGES_REQUESTED** [review]

**Review by MichalMyskow — DISMISSED** [review]

File: src/Http/Requests/Account/UpdateUserPasswordRequest.php [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+use Illuminate\Foundation\Http\FormRequest;
+
+class UpdateUserPasswordRequest extends FormRequest
+{
+    public function rules(): array
+    {
+        return [
+            "password" => ["required", "string", "min:8", "current_password"],
+            "newPassword" => ["required", "string", "min:8", "confirmed"],
+            "newPassword_confirmation" => ["required", "string", "min:8"],
```
Comment (kamilpiech97) [inline src/Http/Requests/Account/UpdateUserPasswordRequest.php:16]: We are using `camelSnake_case`?

File: src/Http/Requests/Account/UpdateUserDataRequest.php (hunk 6 above)
Comment (kamilpiech97) [inline src/Http/Requests/Account/UpdateUserDataRequest.php:18]:
```suggestion
    {
        $userId = $this->user()->id;
        return [
            "email" => ["required", "string", "email", "max:255", "unique:users,email,{$userId},id"],
```

Comment (kamilpiech97) [reply src/Http/Requests/Account/UpdateUserDataRequest.php:18]: It is shorter type of that rule but u can decide which choose.

File: src/Http/Requests/Account/UpdateUserPasswordRequest.php (hunk 10 above)
Comment (kamilpiech97) [reply src/Http/Requests/Account/UpdateUserPasswordRequest.php:16]: Please change this :)

**Review by kamilpiech97 — CHANGES_REQUESTED** [review]

File: src/Http/Requests/Account/UpdateUserPasswordRequest.php (hunk 10 above)
Comment (krzysztofrewak) [reply src/Http/Requests/Account/UpdateUserPasswordRequest.php:16]: I'm afraid that's a Laravel convention.

Comment (Majkulnn) [reply src/Http/Requests/Account/UpdateUserPasswordRequest.php:16]: unfortunately it's laravel convention 
https://laravel.com/docs/9.x/validation#rule-confirmed

File: src/Http/Requests/Account/UpdateUserDataRequest.php (hunk 6 above)
Comment (Majkulnn) [reply src/Http/Requests/Account/UpdateUserDataRequest.php]: i was looking for something like this but there was this version in the documentation

File: src/Http/Requests/Account/UpdateUserPasswordRequest.php (hunk 10 above)
Comment (kamilpiech97) [reply src/Http/Requests/Account/UpdateUserPasswordRequest.php:16]: but something like that `new_password` and `new_password_confirmation` will be better, I think.

**Review by kamilpiech97 — APPROVED** [review]

**Review by MichalMyskow — APPROVED** [review]

**Pull request merged**